File: pfe-java/app/controllers/Authentication.java
package controllers;

import play.data.Form;
import play.data.validation.Constraints;
import play.i18n.Messages;
import play.libs.F;
import play.mvc.Http;
import play.mvc.Result;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class Authentication extends Controller {

    @Inject
    public Authentication(Service service) {
        super(service);
    }

    public static class Login {
        @Constraints.Required
        public String username;
        @Constraints.Required
        public String password;
    }

    static final String USER_KEY = "username";

    public Result login(String returnTo) {
        return ok(views.html.login.render(Form.form(Login.class), returnTo));
    }

    public Result authenticate(String returnTo) {
        Form<Login> submission = Form.form(Login.class).bindFromRequest();
        if (submission.hasErrors()) {
            return badRequest(views.html.login.render(submission, returnTo));
        } else {
            Login login = submission.get();
            if (service.users.authenticate(login.username, login.password)) {
                session().put(USER_KEY, login.username);
                return redirect(returnTo);
            } else {
                submission.reject(Messages.get("auth.unknown", login.username));
                return badRequest(views.html.login.render(submission, returnTo));
            }
        }
    }

    public Result logout() {
        session().remove(USER_KEY);
        return redirect(routes.Items.list());
    }

    public static <A> A authenticated(Http.Context ctx, F.Function<String, A> f, F.Function0<A> g) throws Throwable {
        String username = ctx.session().get(USER_KEY);
        if (username != null) {
            return f.apply(username);
        } else {
            return g.apply();
        }
    }

}

File: pfe-scala/app/controllers/Authentication.scala
package controllers

import javax.inject.{Inject, Singleton}

import play.api.mvc._
import play.api.mvc.Results.Redirect
import play.api.data.Form
import play.api.data.Forms.{tuple, nonEmptyText}
import play.api.i18n.Messages
import scala.concurrent.Future

@Singleton class Authentication @Inject() (service: Service) extends Controller(service) {

  import Authentication._

  def login(returnTo: String) = Action { implicit request =>
    Ok(views.html.login(Login.form, returnTo))
  }

  def authenticate(returnTo: String) = Action { implicit request =>
    val submission = Login.form.bindFromRequest()
    submission.fold(
      errors => BadRequest(views.html.login(errors, returnTo)),
      {
        case (username, password) =>
          if (service.users.authenticate(username, password)) {
            Redirect(returnTo).addingToSession(UserKey -> username)
          } else {
            BadRequest(views.html.login(submission.withGlobalError(Messages("auth.unknown", username)), returnTo))
          }
      }
    )
  }

  val logout = Action { implicit request =>
    Redirect(routes.Items.list()).removingFromSession(UserKey)
  }


}

object Authentication {

  val UserKey = "username"

  type Login = (String, String)
  object Login {
    val form = Form(tuple(
      "username" -> nonEmptyText,
      "password" -> nonEmptyText
    ))
  }

  def authenticated[A](f: String => A, g: => A)(implicit request: RequestHeader): A =
    request.session.get(UserKey) match {
      case Some(username) => f(username)
      case None => g
    }
}

class AuthenticatedRequest[A](val username: String, request: Request[A]) extends WrappedRequest[A](request)

object AuthenticatedAction extends ActionBuilder[AuthenticatedRequest] {
  def invokeBlock[A](request: Request[A], block: AuthenticatedRequest[A] => Future[Result]) =
    Authentication.authenticated(
      username => block(new AuthenticatedRequest(username, request)),
      Future.successful(Redirect(routes.Authentication.login(request.uri)))
    )(request)
}
